src/weights.jl

```diff
@@ -2,44 +2,191 @@
+    s = w.sum
+
+    if corrected
+        sum_sn = zero(eltype(w)) / one(typeof(s)) ^ 2  # to ensure type stability
+        @inbounds for x in w
+            sum_sn += (x / s) ^ 2
+        end
+
+        1 / (s * (1 - sum_sn))
+    else
+        1 / s
+    end
 end
 
+@weights FrequencyWeights
+
 """
-    weights(vs)
+    FrequencyWeights(vs, wsum=sum(vs))
+
+Construct a `FrequencyWeights` vector with weight values `vs`.
+A precomputed sum may be provided as `wsum`.
 
-Construct a `WeightVec` from a given array.
+Frequency weights describe the number of times (or frequency) each observation
+was observed. These weight may also be referred to as case weights or repeat weights.
 """
-weights(vs::RealVector) = WeightVec(vs)
-weights(vs::RealArray) = WeightVec(vec(vs))
+FrequencyWeights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs)) =
+    FrequencyWeights{S, eltype(vs), V}(vs, s)
 
-eltype(wv::WeightVec) = eltype(wv.values)
-length(wv::WeightVec) = length(wv.values)
-values(wv::WeightVec) = wv.values
-sum(wv::WeightVec) = wv.sum
-isempty(wv::WeightVec) = isempty(wv.values)
+"""
+    fweights(vs)
 
-Base.getindex(wv::WeightVec, i) = getindex(wv.values, i)
-Base.size(wv::WeightVec) = size(wv.values)
+Construct a `FrequencyWeights` vector from a given array.
+See the documentation for `FrequencyWeights` for more details.
+"""
+fweights(vs::RealVector) = FrequencyWeights(vs)
+fweights(vs::RealArray) = FrequencyWeights(vec(vs))
 
+"""
+    varcorrection(w::FrequencyWeights, corrected=false)
+
+``\\frac{1}{\sum{w} - 1}``
+"""
+function varcorrection(w::FrequencyWeights, corrected::Bool=false)
+    s = w.sum
+
+    if corrected
+        1 / (s - 1)
+    else
+        1 / s
+    end
+end
+
+@weights ProbabilityWeights
+
+"""
+    ProbabilityWeights(vs, wsum=sum(vs))
+
+Construct a `ProbabilityWeights` vector with weight values `vs`.
+A precomputed sum may be provided as `wsum`.
+
+Probability weights represent the inverse of the sampling probability for each observation,
+providing a correction mechanism for under- or over-sampling certain population groups.
+These weights may also be referred to as sampling weights.
+"""
+ProbabilityWeights{S<:Real, V<:RealVector}(vs::V, s::S=sum(vs)) =
+    ProbabilityWeights{S, eltype(vs), V}(vs, s)
+
+"""
+    pweights(vs)
+
+Construct a `ProbabilityWeights` vector from a given array.
+See the documentation for `ProbabilityWeights` for more details.
+"""
+pweights(vs::RealVector) = ProbabilityWeights(vs)
+pweights(vs::RealArray) = ProbabilityWeights(vec(vs))
+
+"""
+    varcorrection(w::ProbabilityWeights, corrected=false)
+
+``\\frac{n}{(n - 1) \sum w}`` where `n = length(w)`
+"""
+function varcorrection(w::ProbabilityWeights, corrected::Bool=false)
+    s = w.sum
+
+    if corrected
+        n = length(w)
+        n / (s * (n - 1))
+    else
+        1 / s
+    end
+end
+
+"""
+    eweights(n, λ)
+
+Construct an `AnalyticWeights` vector with length `n`,
```

Comment:
I don't think that would be correct either, as the assumptions are also very strong. The `n-1` correction doesn't apply for exponential weights either.